modality-probe-capi/README.md

```diff
@@ -1,53 +1,172 @@
+`header-gen` which you can read more about [here TODO(link)]. The
+short of it is that you’ll need to run the header-gen tool against
+your codebase before it can be compiled because Modality Probe
+actually generates the definitions to those symbols.
+
+```c
+#define DEFAULT_PROBE_SIZE = 7000;
+modality_probe * probe_g = MODALITY_PROBE_NULL_INITIALIZER;
+
+void twist(double x, double y, double z) {
+    int result = MODALITY_PROBE_RECORD(
+        probe_g,
+        TWISTED,
+        "tags=actuation",
+        "A twist command was received"
+    );
+// …
+}
+
+int main() {
+    uint8_t * destination = (uint8_t*)malloc(DEFAULT_PROBE_SIZE);
+    modality_probe_error result = modality_probe_initialize(
+        destination,
+        DEFAULT_PROBE_SIZE,
+        CONTROLLER,
+        probe_g
+    );
+}
+```
+### Interacting with Snapshots
+
+For Modality Probe to do its best work regarding tracking causal
+relationships between components in your system, it needs to be
+included in those component’s interactions. The probes in your system
+derive those relationships by exchanging “points in time” from their
+own logical clocks. It’s through the snapshot API that this exchange
+happens in practice.
+
+``` c
+modality_causal_snapshot snap;
+result = modality_probe_produce_snapshot(CONTROLLER, &snap);
+```
 
-In order to build this library for other platforms, we recommend the use
-of [cargo-xbuild](https://github.com/rust-osdev/cargo-xbuild).
+A snapshot can then be merged into another probe’s log with `merge`:
 
-See that project for detailed instructions. The likely default usage
-path for obtaining a new target's toolchain and building for it is:
+```c
+result = modality_probe_merge_snapshot(ACTUATOR, &snap);
+```
 
-```shell script
-rustup update
-cargo install cargo-xbuild --force
-rustup target add thumbv7em-none-eabi
-rustc -Z unstable-options --print target-spec-json --target thumbv7em-none-eabi | tee thumbv7em-none-eabi.json
-cargo xbuild --target thumbv7em-none-eabi.json
+These things work best when they’re used in-band. That is, they’re
+tacked onto an existing interaction as an extra few bytes that can be
+unpacked on the other end and interpreted as a snapshot that the
+receiving probe can merge into its probe. When these things occur out
+of band, the veracity of the causal relationships Modality Probe is
+meant to capture erodes—the exchanges tell us only that two components
+are related, but not how they are.
+
+### “Now” (Integrating a Probe’s data into your existing logging)
+
+Modality Probe’s “now” API allows one to associate traced events with
+your existing logging infrastructure by embedding a probe’s current
+clock value into your log messages:
+
+```c
+#ifdef MODALITY_PROBE_MACROS_ENABLED
+    modality_probe_instant now = modality_probe_now(CONTROLLER);
+    LOG(
+        “controller: sent a twist command, probe clock (%d, %d)”,
+        now.clock.id,
+        now.clock.count
```

Comment:
The new `modality_probe_instant` type has changed a bit, probably ought to be (`now.clock.id`, `now.clock.epoch`, `now.clock.ticks`, `now.event_count`).